utils/include/utils/bk_tree.h

```diff
@@ -0,0 +1,144 @@
+
+#include <algorithm>
+#include <array>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hlasm_plugin::utils {
+
+template<typename T, class Distance>
+class bk_tree
+{
+    [[no_unique_address]] Distance m_dist;
+
+    static constexpr size_t invalid = (size_t)-1;
+
+    struct node_t
+    {
+        size_t distance;
+        size_t next_sibling;
+        size_t first_child;
+        T value;
+    };
+
+    std::vector<node_t> m_nodes;
+
+    static constexpr auto dist = [](size_t a, size_t b) {
+        if (a >= b)
+            return a - b;
+        else
+            return b - a;
+    };
+
+    template<size_t result_size, typename U>
+    void find_impl(std::array<std::pair<const T*, size_t>, result_size>& result,
+        const U& value,
+        size_t node_id,
+        size_t distance) const noexcept(noexcept(m_dist(std::declval<const T&>(), value)))
+    {
+        auto& best = result.front();
+        while (node_id != invalid && best.second > 0)
+        {
+            const auto& node = m_nodes[node_id];
+            node_id = node.next_sibling;
+
+            if (dist(node.distance, distance) > best.second)
+                continue;
+
+            const auto d = m_dist(node.value, value);
+            if (d <= best.second)
+            {
+                std::shift_right(result.begin(), result.end(), 1);
+                best = { &node.value, d };
+            }
+
+            if (d == 0)
+                break;
+
+            if (node.first_child != invalid)
+                find_impl(result, value, node.first_child, d);
+        }
+    }
+
+public:
+    bk_tree() requires(std::is_default_constructible_v<Distance>) = default;
+    explicit bk_tree(Distance d) noexcept(std::is_nothrow_move_constructible_v<Distance>)
+        : m_dist(std::move(d))
+    {}
+
+    auto size() const noexcept { return m_nodes.size(); }
+    void clear() noexcept { m_nodes.clear(); }
+
+    template<typename U>
+    std::pair<const T*, bool> insert(U&& value)
+    {
+        size_t dummy = m_nodes.empty() ? invalid : 0;
+        size_t distance = invalid;
+        auto* node_id = &dummy;
+        while (*node_id != invalid)
+        {
+            auto& node = m_nodes[*node_id];
+            if (node.distance != distance)
+            {
+                node_id = &node.next_sibling;
+                continue;
+            }
+            node_id = &node.first_child;
+            distance = m_dist(std::as_const(node.value), std::as_const(value));
+            if (distance == 0)
+                return { &node.value, false };
+        }
+        struct restore_t
+        {
+            size_t* id;
+            ~restore_t()
+            {
+                if (id)
+                    *id = invalid;
+            }
+        } restore { node_id };
+
+        *node_id = m_nodes.size();
+        const auto& new_node = m_nodes.emplace_back(node_t { distance, invalid, invalid, std::forward<U>(value) });
```

Comment:
This comes from SonarCloud
```suggestion
        const auto& new_node = m_nodes.emplace_back(distance, invalid, invalid, std::forward<U>(value));
```

Reply:
no go - clang 12-14 issue

Reply:
Right...